pkg/broker/broker.go

```diff
@@ -295,67 +309,96 @@ func (a AnsibleBroker) Provision(instanceUUID uuid.UUID, req *ProvisionRequest,
 	//    https://github.com/openservicebrokerapi/servicebroker/issues/127
 	// -> Atomic deprovision and removal of service entry in etcd?
 	//    * broker::Deprovision
 	//    Arguments for this? What data do apbs require to deprovision?
+	//    * namespace
 	//    Maybe just hand off a serialized ServiceInstance and let the apb
 	//    decide what's important?
+	//    * delete credentials from etcd
 	//    * if noerror: delete serviceInstance entry with Dao
-	////////////////////////////////////////////////////////////
-	instanceID := instanceUUID.String()
-
-	if err := a.validateDeprovision(instanceID); err != nil {
+	instance, err := a.getServiceInstance(instanceUUID)
+	if err != nil {
 		return nil, err
 	}
-	instance, err := a.dao.GetServiceInstance(instanceID)
-	/// Handle case where service is not found in etcd
-	if client.IsKeyNotFound(err) {
-		a.log.Debug("unable to find service instance - %#v", err)
-		return nil, ErrorNotFound
-	}
-	// bubble up  error
-	if err != nil {
-		a.log.Error("Error from etcd - %#v", err)
+
+	if err := a.validateDeprovision(instance); err != nil {
 		return nil, err
 	}
 
-	podName, err := apb.Deprovision(instance, a.log)
+	var token string
+
+	if async {
+		a.log.Info("ASYNC deprovision in progress")
+		// asynchronously provision and return the token for the lastoperation
+		dpjob := NewDeprovisionJob(instance, a.clusterConfig, a.dao, a.log)
+
+		token = a.engine.StartNewJob(dpjob)
+
+		// HACK: there might be a delay between the first time the state in etcd
+		// is set and the job was already started. But I need the token.
+		a.dao.SetState(instanceUUID.String(), apb.JobState{Token: token, State: apb.StateInProgress})
+		return &DeprovisionResponse{Operation: token}, nil
+	} else {
+		// TODO: do we want to do synchronous deprovisioning?
+		a.log.Info("Synchronous deprovision in progress")
+		podName, err := apb.Deprovision(instance, a.clusterConfig, a.log)
+		err = cleanupDeprovision(err, podName, instance, a.dao, a.log)
+		if err != nil {
+			return nil, err
+		}
+		return &DeprovisionResponse{}, nil
+	}
+}
 
-	sm := apb.NewServiceAccountManager(a.log)
-	a.log.Info("Destroying APB sandbox...")
+func cleanupDeprovision(err error, podName string, instance *apb.ServiceInstance, dao *dao.Dao, log *logging.Logger) error {
+	instanceID := instance.Id.String()
+	sm := apb.NewServiceAccountManager(log)
+	log.Info("Destroying APB sandbox...")
 	sm.DestroyApbSandbox(podName, instance.Context.Namespace)
 
 	if err == docker.ErrNoSuchImage {
-		a.log.Debug("unable to find service instance - %#v", err)
-		return nil, ErrorNotFound
+		log.Debug("unable to find service instance - %#v", err)
+		return ErrorNotFound
 	}
 	// bubble up error.
 	if err != nil {
-		a.log.Error("error from deprovision - %#v", err)
-		return nil, err
+		log.Error("error from deprovision - %#v", err)
+		return err
 	}
 
-	a.dao.DeleteServiceInstance(instanceID)
+	if err := dao.DeleteExtractedCredentials(instanceID); err != nil {
```

Comment:
What we could do is have a sandbox cleanup function and pull 382 - 390 back into deprovision.